### PR TITLE
Modified the Coefficients Table to be instantiated with dictionaries as well as strings

### PR DESCRIPTION
**Pull request by g-weatherill**
It is increasingly common to modify coefficients of published GMPEs depending on the application. This adapts the Coefficients Table object to support input of the coefficients in terms of a tuple of dictionaries (one for SA coefficients the other for Non-SA coefficients), in addition to the current means of definition using a string. This can allow for the coefficients table to be modified dynamically without the need for multiple tables to be hard coded. This is needed for https://github.com/gem/oq-engine/issues/2954

### CODE REVIEW

**Comment by micheles** [comment]
Why force the user to distinguish SA coefficients and non-SA coefficients? The user could pass a single dictionary and inside the code could distinguish if the IMT is a SA one or not.

File: openquake/hazardlib/tests/gsim/base_test.py [hunk 1]
```diff
@@ -634,3 +635,49 @@ def test_ordering_and_equality(self):
...
+            table1.sa_coeffs,
+            {SA(period=0.1, damping=5): {"a": 1.0, "b": 2.0},
+             SA(period=1.0, damping=5): {"a": 5.0, "b": 10.0},
+             SA(period=10.0, damping=5): {"a": 10.0, "b": 20.0}}
+            )
+
+    def test_table_dict_instantiation(self):
+        # Check that the table instantiates with pre-defined dictionaries
+        table1 = CoeffsTable(sa_damping=5, table=self.coefficient_string)
+        # Create target dictionatu
```
Comment (mmpagani) [inline openquake/hazardlib/tests/gsim/base_test.py]: dictionary

File: openquake/hazardlib/gsim/base.py [hunk 2]
```diff
@@ -912,20 +912,46 @@ class CoeffsTable(object):
...
             raise TypeError('CoeffsTable requires "table" kwarg')
-        table = kwargs.pop('table').strip().splitlines()
+        table = kwargs.pop('table')
+        self.sa_coeffs = {}
+        self.non_sa_coeffs = {}
         sa_damping = kwargs.pop('sa_damping', None)
         if kwargs:
             raise TypeError('CoeffsTable got unexpected kwargs: %r' % kwargs)
+        if isinstance(table, str):
+            self._setup_table_from_str(table, sa_damping)
```
Comment (micheles) [inline openquake/hazardlib/gsim/base.py]: Did you run the tests both with Python 2 and Python 3? I wonder if the str/bytes change may cause trouble here.

Comment (daniviga) [inline openquake/hazardlib/gsim/base.py]:
```python
test_table_bad_instantiation (openquake.hazardlib.tests.gsim.base_test.CoeffsTableTestCase) ... FAIL

======================================================================
FAIL: test_table_bad_instantiation (openquake.hazardlib.tests.gsim.base_test.CoeffsTableTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dvigano/openquake/oq-engine/openquake/hazardlib/tests/gsim/base_test.py", line 682, in test_table_bad_instantiation
    "CoeffsTable cannot be constructed with "
AssertionError: "CoeffsTable cannot be constructed with inputs of the form <class 'int'>" != "CoeffsTable cannot be constructed with inputs of the form <type 'int'>"
- CoeffsTable cannot be constructed with inputs of the form <class 'int'>
?                                                            ^^^^^
+ CoeffsTable cannot be constructed with inputs of the form <type 'int'>
?                                                            ^^^^


----------------------------------------------------------------------
Ran 528 tests in 78.733s
```

**Comment by daniviga** [comment]
Hazardlib tests on python3 are broken:

```bash
test_table_bad_instantiation (openquake.hazardlib.tests.gsim.base_test.CoeffsTableTestCase) ... FAIL

======================================================================
FAIL: test_table_bad_instantiation (openquake.hazardlib.tests.gsim.base_test.CoeffsTableTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dvigano/openquake/oq-engine/openquake/hazardlib/tests/gsim/base_test.py", line 682, in test_table_bad_instantiation
    "CoeffsTable cannot be constructed with "
AssertionError: "CoeffsTable cannot be constructed with inputs of the form <class 'int'>" != "CoeffsTable cannot be constructed with inputs of the form <type 'int'>"
- CoeffsTable cannot be constructed with inputs of the form <class 'int'>
?                                                            ^^^^^
+ CoeffsTable cannot be constructed with inputs of the form <type 'int'>
?                                                            ^^^^
```

File: debian/changelog_hazardlib [hunk 3]
```diff
@@ -1,3 +1,6 @@
+  [Graeme Weatherill]
+  * Adapts CoeffsTable to be instantiated with dictionarys as well as strings
+
```
Comment (mmpagani) [inline debian/changelog_hazardlib]: Move this to the 'changelog' file. See https://github.com/gem/oq-engine/blob/master/debian/changelog

File: openquake/hazardlib/gsim/base.py (hunk 2 above)
Comment (g-weatherill) [inline openquake/hazardlib/gsim/base.py]: What is the python 2 and 3 compatible way to do this?

Also, the error Daniele notes is due to the use of `type(foo)` is giving a different string between Python 2 and 3. Is `x.__class__.__name__` safer/cleaner to use?

**Comment by micheles** [comment]
LGTM

**Pull request merged**